src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/coding/illegaltype/InputIllegalTypeWithRecordPattern.java

```diff
@@ -0,0 +1,87 @@
+/*
+IllegalType
+validateAbstractClassNames = (default)false
+illegalClassNames = HashMap, HashSet, LinkedHashMap, LinkedHashSet, TreeMap, TreeSet, \
+                    java.util.HashMap, java.util.HashSet, java.util.LinkedHashMap, \
+                    java.util.LinkedHashSet, java.util.TreeMap, java.util.TreeSet, \
+                    Rectangle, ColoredPoint, Point, Box
+legalAbstractClassNames = (default)
+ignoredMethodNames = (default)getEnvironment, getInitialContext
+illegalAbstractClassNameFormat = (default)^(.*[.])?Abstract.*$
+memberModifiers = (default)
+tokens = (default)ANNOTATION_FIELD_DEF, CLASS_DEF, INTERFACE_DEF, METHOD_CALL, METHOD_DEF, \
+         METHOD_REF, PARAMETER_DEF, VARIABLE_DEF, PATTERN_VARIABLE_DEF, RECORD_DEF, \
+         RECORD_COMPONENT_DEF, RECORD_PATTERN_DEF
+
+
+*/
+
+//non-compiled with javac: Compilable with Java21
+package com.puppycrawl.tools.checkstyle.checks.coding.illegaltype;
+
+import java.util.ArrayList;
+import java.util.LinkedHashMap;
+import java.util.List;
+
+public class InputIllegalTypeWithRecordPattern {
+
+    record ColoredPoint(Point p, String c) { } // violation, 'Usage of type 'Point' is not allowed.'
+    record Rectangle(ColoredPoint upperLeft, ColoredPoint lowerRight) { }
+    // 2 violations above:
+    //                    'Usage of type 'ColoredPoint' is not allowed.'
+    //                    'Usage of type 'ColoredPoint' is not allowed.'
+    record Point(int x, int y) { }
+    record Box<T>(T t) { }
+
+
+    public void testWithInstanceOf(Object obj) {
+        // violation below, 'Usage of type 'Rectangle' is not allowed.'
+        if (obj instanceof Rectangle r) { }
+        if (obj instanceof Rectangle(ColoredPoint x, ColoredPoint y)) { }
+        // 3 violations above:
+        //                  'Usage of type 'Rectangle' is not allowed.'
+        //                  'Usage of type 'ColoredPoint' is not allowed.'
+        //                  'Usage of type 'ColoredPoint' is not allowed.'
+        if (obj instanceof Rectangle(ColoredPoint(Point(int _,int _),_),_)) { }
+        // 3 violations above:
+        //                  'Usage of type 'Rectangle' is not allowed.'
+        //                  'Usage of type 'ColoredPoint' is not allowed.'
+        //                  'Usage of type 'Point' is not allowed.'
+        if (obj instanceof Point(_,_)) { }
+        // violation above, 'Usage of type 'Point' is not allowed.'
+
+    }
+
+    public void testWithInstanceOfWithGenerics() {
+
+        // violation below, 'Usage of type 'LinkedHashMap' is not allowed.'
+        LinkedHashMap<Integer, Integer> l2
+                = new LinkedHashMap<>();
+
+        Box<LinkedHashMap<Integer,Integer>> box = new Box<>(l2);
```

Comment:
Guess this is more me being curious. Since we are changing pattern defs, it would make things easier to only see pattern def violations in this file. This is what lead me to my next curious question.

Would changing these 2 statements to the following, and leaving the rest of the code the same still compile?
````
        Map<Integer, Integer> l2 = new LinkedHashMap<>();

        Box<Map<Integer,Integer>> box = new Box<>(l2);
````
Would that mean we could have different type checks:
````
if (box instanceof Box<LinkedHashMap<Integer,Integer>>(var linkedHashMap)) {}

if (box instanceof Box<HashMap<Integer,Integer>>(var hashMap)) {}
````

Reply:
> Would changing these 2 statements to the following, and leaving the rest of the code the same still compile?
> 
> ```
>         Map<Integer, Integer> l2 = new LinkedHashMap<>();
> 
>         Box<Map<Integer,Integer>> box = new Box<>(l2);
> ```

Nope. 

`Inconvertible types; cannot cast 'com.puppycrawl.tools.checkstyle.checks.coding.illegaltype.InputIllegalTypeWithRecordPattern.Box<java.util.Map<java.lang.Integer,java.lang.Integer>>' to 'com.puppycrawl.tools.checkstyle.checks.coding.illegaltype.InputIllegalTypeWithRecordPattern.Box<java.util.LinkedHashMap<java.lang.Integer,java.lang.Integer>>'`